### PR TITLE
Using core::Optional to initialize optional variables

### PR DESCRIPTION
**Pull request by Villaquiranm**
Hello @gavv ,

I started working in this and following the use of core::Optional that is made on `src/modules/roc_ctl/task_queue.h`

So I change the declaration and the initialization of some variables without problems, (I execute the Travis build to check every file changed).

When i arrived to `receiver_session.h` it seems like when changing the line 99 **fec_decoder_** from _ScopedPtr_ to _Optional_ we got a core dumped.

So maybe I've changed a variable that was not supposed to be changed ? how could i identify them ? Before that last variable this check was working fine.

```
     TEST   roc_core
..................................................
.......................
OK (73 tests, 73 ran, 1036 checks, 0 ignored, 0 filtered out, 9 ms)

     TEST   roc_address
...........................................
OK (43 tests, 43 ran, 553 checks, 0 ignored, 0 filtered out, 1 ms)

     TEST   roc_packet
.....................................
OK (37 tests, 37 ran, 1289 checks, 0 ignored, 0 filtered out, 1 ms)

     TEST   roc_audio
..................................................
...........................
OK (77 tests, 77 ran, 328443 checks, 0 ignored, 0 filtered out, 944 ms)

     TEST   roc_rtp
................
OK (16 tests, 16 ran, 1571 checks, 0 ignored, 0 filtered out, 0 ms)

     TEST   roc_fec
..................................................
.
OK (51 tests, 51 ran, 1598028 checks, 0 ignored, 0 filtered out, 464 ms)

     TEST   roc_netio
........................
OK (24 tests, 24 ran, 18310 checks, 0 ignored, 0 filtered out, 107 ms)

     TEST   roc_sndio
...............
OK (15 tests, 15 ran, 10350 checks, 0 ignored, 0 filtered out, 60 ms)

     TEST   roc_pipeline
.....................
src/modules/roc_core/optional.h:44: error: roc_panic()

ERROR: roc_pipeline: optional: attempt to set incorrect object

Backtrace:
#1: 0x561780b9e519 roc::core::crash(char const*)+0x39
#2: 0x561780b9da81 roc::core::panic(char const*, char const*, int, char const*, ...)+0x131
#3: 0x561780b6ddf9 _ZN3roc8pipeline15ReceiverSessionC1ERKNS0_21ReceiverSessionConfigERKNS0_20ReceiverCommonConfigERKNS+0x0
#4: 0x561780b6e5e0 _ZN3roc8pipeline20ReceiverSessionGroup15create_session_ERKNS_4core9SharedPtrINS_6packet6PacketENS2_+0x0
#5: 0x561780b7491a roc::pipeline::ReceiverEndpoint::flush_packets()+0x12a
#6: 0x561780b6c9e0 roc::pipeline::ReceiverEndpointSet::update(unsigned int)+0x20
#7: 0x561780b6f3c0 roc::pipeline::ReceiverSource::process_frame_imp(roc::audio::Frame&)+0x70
#8: 0x561780b731a3 roc::pipeline::TaskPipeline::process_next_subframe_(roc::audio::Frame&, unsigned long*)+0x73
#9: 0x561780b73737 roc::pipeline::TaskPipeline::process_frame_and_tasks_precise_(roc::audio::Frame&)+0x97
#10: 0x561780b6f26d roc::pipeline::ReceiverSource::read(roc::audio::Frame&)+0x4d
#11: 0x561780b555f2 roc::pipeline::TEST_GROUP_CppUTestGroupsender_sink_receiver_source::send_receive(int, unsigned long)+0x1872
#12: 0x561780bc1102 PlatformSpecificSetJmpImplementation+0x42
#13: 0x561780bbe52b Utest::run()+0x2b
#14: 0x561780bc0aaf UtestShell::runOneTestInCurrentProcess(TestPlugin*, TestResult&)+0x6f
#15: 0x561780bc1102 PlatformSpecificSetJmpImplementation+0x42
#16: 0x561780bbe637 UtestShell::runOneTest(TestPlugin*, TestResult&)+0x57
#17: 0x561780bbdbe7 TestRegistry::runAllTests(TestResult&)+0xa7
#18: 0x561780bb2e39 CommandLineTestRunner::runAllTests()+0xe9
#19: 0x561780bb2f29 CommandLineTestRunner::runAllTestsMain()+0x79
#20: 0x561780bb308d CommandLineTestRunner::RunAllTests(int, char const**)+0xad
#21: 0x561780b36cf6 main+0xd6
#22: 0x7ff4b676a0b3 __libc_start_main+0xf3
#23: 0x561780b36dce _start+0x2e
Aborted (core dumped)
scons: *** [test/roc_pipeline] Error 134
```

### CODE REVIEW

**Comment by gavv** [comment]
That's right, we can't use `Optional<>` for FEC decoder.

`Optional<>` is intended for use with concrete types. The type and size of the object should be known at compile time. `Optional<T>` has an embedded storage of size `sizeof(T)` and in its destructor it calls `~T()`. The object should be allocated directly on that embedded storage using placement new.

On the other hand, FEC codecs and some other codecs (resampler, audio codecs, maybe something else) are interfaces and the specific implementation is selected at runtime (e.g. in fec::CodecMap by codec id).

Using such objects with `Optional<>` is wrong for two reasons: 1) we don't know the object size at compile time; sizeof(T) would give us the size of the interface (like IBlockEncoder), but not the size of the created object; 2) such objects are explicitly created on heap using IAllocator (e.g. in CodecMap) which always places them in heap.

So just don't use `Optional<>` with objects which type is not known at compile time.

**Comment by gavv** [comment]
#389

File: src/modules/roc_pipeline/converter_sink.h [hunk 1]
```diff
@@ -57,13 +57,13 @@ class ConverterSink : public sndio::ISink, public core::NonCopyable<> {
 private:
     audio::NullWriter null_writer_;
 
-    core::ScopedPtr<audio::PoisonWriter> resampler_poisoner_;
-    core::ScopedPtr<audio::ResamplerWriter> resampler_writer_;
-    core::ScopedPtr<audio::IResampler> resampler_;
+    core::Optional<audio::PoisonWriter> resampler_poisoner_;
+    core::Optional<audio::ResamplerWriter> resampler_writer_;
+    core::Optional<audio::IResampler> resampler_;
```
Comment (Villaquiranm) [inline src/modules/roc_pipeline/converter_sink.h:62]: @gavv Just another question, thanks for you review. 

In this line `core::Optional<audio::IResampler> resampler_;` i should not use **core::Optional** as **audio::IResampler**  interface right ?

Comment (gavv) [reply src/modules/roc_pipeline/converter_sink.h:62]: Yep.

**Comment by Villaquiranm** [comment]
@gavv , I try to follow your suggestions about the use of *core::Optional** class.

in `src/modules/roc_pipeline/receiver_endpoint_set.h` we have this two variables 
```
core::ScopedPtr<ReceiverEndpoint> source_endpoint_;
core::ScopedPtr<ReceiverEndpoint> repair_endpoint_;
```
They use the method **reset()** without arguments so i imagine that with those we should use the **core::ScopedPtr** right ?

**Comment by gavv** [comment]
We can just add support for `reset()` (and `reset(NULL)`) to core::Optional.

**Comment by gavv** [comment]
It seems macos builds are broken on develop again: https://travis-ci.org/github/roc-streaming/roc-toolkit/builds/708683310

Likely scons version changed in brew. Don't have time for a closer look right now.

**Comment by Villaquiranm** [comment]
Hello @gavv , I have implemented the reset method in Optional, what do you think about it ? 

About the error in the pipeline, there is something i can do about it?

**Comment by gavv** [comment]
Cool! I'll be able to review it in a day or two. Regarding macos error, it's unrelated to the PR, if you wish and have time you can fix it and make another PR. Otherwise I'll look at it in upcoming days.

**Comment by gavv** [comment]
macOS travis build is fixed in fresh develop, please rebase.

**Comment by Villaquiranm** [comment]
Hello @gavv the build seems to be working now thanks for the fix.

File: src/modules/roc_core/optional.h [hunk 2]
```diff
@@ -46,6 +46,12 @@ template <class T> class Optional : public NonCopyable<> {
         ptr_ = ptr;
     }
 
+    //! Delete the pointed object and reset the structure
+    void reset() {
```
Comment (gavv) [inline src/modules/roc_core/optional.h:50]: With this addition, the behavior of reset method becomes a bit inconsistent.

This is now legal:

```
foo.reset(some_ptr);
foo.reset();
foo.reset(another_ptr);
```

Bit this is still not:

```
foo.reset(some_ptr);
foo.reset(NULL);
foo.reset(another_ptr);
```

And this is not legal too:

```
foo.reset(some_ptr);
foo.reset(another_ptr);
```

If we're adding support for resetting multiple times, we should adjust the existing reset(T*) version as well and allow all of the examples above.

Actually, after doing it, we don't event need reset() without arguments, we can just use reset(NULL).

Comment (Villaquiranm) [reply src/modules/roc_core/optional.h]: Cool, thanks for your review

I had to add this **ptr &&** in optional.h line 43
```
if (ptr && (void*)ptr != storage_.mem) {
     roc_panic("optional: attempt to set incorrect object");
``
to deal with the situation when we set NULL as pointer and don't panic

File: src/modules/roc_core/optional.h [hunk 3]
```diff
@@ -85,6 +85,11 @@ template <class T> class Optional : public NonCopyable<> {
         MaxAlign align;
         char mem[sizeof(T)];
     };
+    void destroy_() {
```
Comment (gavv) [inline src/modules/roc_core/optional.h:88]: Nitpick: missing empty line.

Also, since destroy_() is short and used only once, I'd inline it into reset().

**Comment by gavv** [comment]
> Build status: **FAILED**
> Build log: https://travis-ci.org/roc-streaming/roc-toolkit/builds/714162846

Yet another issue with macos build. Fixed in develop, please rebase.

File: src/modules/roc_core/optional.h (hunk 3 above)
Comment (Villaquiranm) [reply src/modules/roc_core/optional.h]: Yeah you're right it is better now 👍

**Comment by gavv** [comment]
Thanks!

**Pull request merged**